lightning/src/onion_message/messenger.rs

```diff
@@ -0,0 +1,399 @@
+		control_tlvs_keys.push(control_tlvs_ss);
+
+		let (rho, mu) = onion_utils::gen_rho_mu_from_shared_secret(onion_packet_ss.as_ref());
+		onion_packet_keys.push(onion_utils::OnionKeys {
+			#[cfg(test)]
+			shared_secret: onion_packet_ss,
+			#[cfg(test)]
+			blinding_factor: _blinding_factor,
+			ephemeral_pubkey,
+			rho,
+			mu,
+		});
+	})?;
+
+	Ok((control_tlvs_keys, onion_packet_keys))
+}
+
+/// Build an onion message's payloads for encoding in the onion packet.
+fn build_payloads(intermediate_nodes: Vec<PublicKey>, destination: Destination, mut encrypted_tlvs_keys: Vec<[u8; 32]>) -> Vec<(Payload, [u8; 32])> {
+	let num_intermediate_nodes = intermediate_nodes.len();
+	let num_payloads = num_intermediate_nodes + destination.num_hops();
+	debug_assert_eq!(encrypted_tlvs_keys.len(), num_payloads);
+	let mut payloads = Vec::with_capacity(num_payloads);
+	let mut enc_tlv_keys = encrypted_tlvs_keys.drain(..);
+	for pk in intermediate_nodes.into_iter().skip(1) {
+		payloads.push((Payload::Forward(ForwardControlTlvs::Unblinded(
+						ForwardTlvs {
+							next_node_id: pk,
+							next_blinding_override: None,
+						}
+		)), enc_tlv_keys.next().unwrap()));
+	}
+	match destination {
+		Destination::Node(pk) => {
+			if num_intermediate_nodes != 0 {
+				payloads.push((Payload::Forward(ForwardControlTlvs::Unblinded(
+								ForwardTlvs {
+									next_node_id: pk,
+									next_blinding_override: None,
+								}
+				)), enc_tlv_keys.next().unwrap()));
+			}
+			payloads.push((Payload::Receive { control_tlvs: ReceiveControlTlvs::Unblinded(
+						ReceiveTlvs {
+							path_id: None,
+						}
+			)}, enc_tlv_keys.next().unwrap()));
+		},
+		Destination::BlindedRoute(BlindedRoute { introduction_node_id, blinding_point, blinded_hops }) => {
+			if num_intermediate_nodes != 0 {
+				payloads.push((Payload::Forward(ForwardControlTlvs::Unblinded(
+								ForwardTlvs {
+									next_node_id: introduction_node_id,
+									next_blinding_override: Some(blinding_point),
+								}
+				)), enc_tlv_keys.next().unwrap()));
+			}
+			let num_hops = blinded_hops.len();
+			for (idx, hop) in blinded_hops.into_iter().enumerate() {
+				if idx != num_hops - 1 {
+					payloads.push((Payload::Forward(ForwardControlTlvs::Blinded(hop.encrypted_payload)),
+					enc_tlv_keys.next().unwrap()));
+				} else {
+					payloads.push((Payload::Receive {
+						control_tlvs: ReceiveControlTlvs::Blinded(hop.encrypted_payload),
+					}, enc_tlv_keys.next().unwrap()));
+				}
+			}
+		}
+	}
+	payloads
+}
+
+/// Errors if the serialized payload size exceeds onion_message::BIG_PACKET_HOP_DATA_LEN
+fn construct_onion_message_packet(payloads: Vec<(Payload, [u8; 32])>, onion_keys: Vec<onion_utils::OnionKeys>, prng_seed: [u8; 32]) -> Result<Packet, ()> {
+	let payloads_serialized_len: usize = payloads.iter().map(|p| p.serialized_length() + 32 /* HMAC */).sum();
```

Comment:
Do you mean to add a `payloads_serialized_len()` method to `onion_utils` and call that instead of calculating here? or is there more to it than that?